Create initial sharding playbook

haydentherapper opened the pull request:
This includes the sharding playbook to be used by oncallers to spin up a new shard.

There's a TODO to add a dashboard link once we have monitoring, and WIPs as we are actively developing the prober and figuring out shard turn down.

Towards https://github.com/sigstore/rekor-tiles/issues/40

<!--
Thanks for opening a pull request! Please do not just delete this text.  The three fields below are mandatory.

Please remember to:
- This PR requires an issue. If it is a new feature, the issue should proceed the PR and will have allowed sufficient time for discussions to take place. Please use
issue tags such as "Closes #XYZ" or "Resolves sigstore/repo-name#XYZ".
  [documentation](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- ensure your commits are signed-off, as sigstore uses the [DCO](https://en.wikipedia.org/wiki/Developer_Certificate_of_Origin) using `git commit -s`, or `git commit -s --amend` if you want to amend already existing commits
- lastly, ensure there are no merge commits!

Thank you :)
-->

#### Summary
<!--
 Explain the **motivation** for making this change. What existing problem does the pull request solve? How can reviewers test this PR?
-->

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to an administrator running private sigstore instances (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense.

-->

#### Documentation
<!--

Does this change require an update to documentation? How will users implement your new feature?

Please reference a PR within https://docs.sigstore.dev

-->

loosebazooka commented on SHARDING.md:
```diff
@@ -0,0 +1,254 @@
+# Sharding Playbook
+
+This playbook walks through the steps to create a new log shard.
+The steps should be the same for Rekor and the CT log, though the
+CT process for generating key material may differ.
```
rekor shouldn't know about CT, or this should be in another repo?

haydentherapper replied:
Will update, just wanted to state that the Rekor and CT sharding processes should be the same.

loosebazooka commented on SHARDING.md:
```diff
@@ -0,0 +1,254 @@
+
+Roughly, the steps are:
+
+1. Create the GCP resources with Terraform
```
same here: this seems specific to our public good instance.

haydentherapper replied:
Yep, I'll move this playbook to PGI.

loosebazooka commented on SHARDING.md:
```diff
@@ -0,0 +1,254 @@
+To add a new shard, create a module under either
+`terraform/environments/staging/1-infrastructure/staging.tf` or
+`terraform/environments/production/1-infrastructure/production.tf`.
+Name the module `rekor-tiles-log-<year>-<revision>`, e.g.
```
what does revision mean here? That we can have multiple shards per year? Should they always count up?

nvm, somewhat answered in the example below, but some guidance on revision would still be helpful.

loosebazooka commented on SHARDING.md:
```diff
@@ -0,0 +1,254 @@
+## Create GCP resources
+
+New shards are declared as additional modules under the respective
+environment. This is a different pattern than other modules, which are all
+initialized through the `sigstore` module. Declaring each shard as its
```
I'm not sure exactly what this means (probably because I have no context into infra). Why are other modules initialized through the single `sigstore` module?

loosebazooka commented on SHARDING.md:
```diff
@@ -0,0 +1,254 @@
+concurrently updated.
+
+To add a new shard, create a module under either
+`terraform/environments/staging/1-infrastructure/staging.tf` or
+`terraform/environments/production/1-infrastructure/production.tf`.
```
can these be links to somewhere?

loosebazooka commented on SHARDING.md:
```diff
@@ -0,0 +1,254 @@
+
+To create a shard signing key:
+
+1. Grant yourself access to use the new signing key. Update the `staging.tf`
```
you can just do this without someone approving it?

loosebazooka commented on SHARDING.md:
```diff
@@ -0,0 +1,254 @@
+
+<!--- TODO: Add code pointers for values -->
+
+Merge, and wait for ArgoCD to create the resources and spin up the service.
```
wait how long? is there some indicator? maybe an an oncall person would already know this?

loosebazooka commented on SHARDING.md:
```diff
@@ -0,0 +1,254 @@
+
+## Update Prober
+
+Update the prober configuration for
```
does prober work with v2?

haydentherapper replied:
Tracking in https://github.com/sigstore/rekor-tiles/issues/46

loosebazooka commented on SHARDING.md:
```diff
@@ -0,0 +1,254 @@
+Using Cosign, test signing and verification against this new shard. You will need to manually update the SigningConfig
+and TrustedRoot files and pass them via CLI flags since those files are pulled from TUF and will not yet have been updated.
+
+<!--- TODO: Update with example TrustedRoot and SigningConfig -->
```
include example call to cosign using custom trustedroot and signingconfig.

haydentherapper replied:
Updated with a curl-based example rather than cosign.

loosebazooka commented on SHARDING.md:
```diff
@@ -0,0 +1,254 @@
+of one week, this means setting the validity window to be at least one week
+past when the TUF targets are signed.
+
+You will need to estimate two timestamps:
```
I think we should have a well defined overlap period. maybe 1 day/week/month or whatever?

Having to guess this means people will ask around to get a recommendation instead of us just having this decision already made.

haydentherapper commented on SHARDING.md:
```diff
@@ -0,0 +1,254 @@
+
+<!--- TODO: Specify how we should do this -->
+
+<!--- TODO: Specify how to convey to monitors the shard is frozen.
```
One other thought - For CT, we can't add extension lines, so we if we want to denote a frozen shard in the same way, baking in the checkpoint might be the only viable approach.

haydentherapper reviewed (commented):
Thanks for the feedback, will work on updating!

jku commented on SHARDING.md:
```diff
@@ -0,0 +1,254 @@
+
+You will need to estimate two timestamps:
+
+* New shard's start for SigningConfig, which should be one week past when the TUF
+timestamp will be signed. Overestimate if needed, as you don't want clients to start
```
Would this be more understandable to folks with no knowledge of root-signing details?

```suggestion
* New shard's start for SigningConfig, which should be one week past when the root-signing PR is merged and new TUF repository is published. Overestimate if needed, as you don't want clients to start
```

jku commented on SHARDING.md:
```diff
@@ -0,0 +1,254 @@
+* New shard's start for SigningConfig, which should be one week past when the TUF
+timestamp will be signed. Overestimate if needed, as you don't want clients to start
+using the new shard before all clients have picked up the latest TrustedRoot.
+* Previous shard's end for TrustedRoot, which is when the previous shard's write path
+is turned off. This must be after the start of the new shard. It doesn't have to
```
```suggestion
* Previous shard's end for TrustedRoot: clients will consider this shards entry timestamps invalid if they are after this date. This must be after the start of the new shard. It doesn't have to
```

I don't know if the suggestion is any better -- you could reword the original as well to make it a bit clearer (that this date doesn't need to be the moment shard is turned off, just needs to be later than the turn off date.

jku commented on SHARDING.md:
```diff
@@ -0,0 +1,254 @@
+        "keyDetails": "...",
+        "validFor": {
+          "start": "...",
+          "end": "<UTC timestamp for approx when the shard will be shut down>"
```
```suggestion
          "end": "<UTC timestamp that is after the shard is turned read only>"
```

jku reviewed (commented):
This makes sense to me.

haydentherapper commented:
This is ready to go. It's been updated with example PRs based on when @cmurphy and I turned up the first shard. Everything under `## Steps` and `## Verify Shard Health` should be complete and up-to-date.

There's two WIPs:

* Probers. Shard discovery should be automatic, both for the write path with the SigningConfig and read path with the TrustedRoot. I added a TODO to the prober tracking issue to allow for us to manually set the shard URL in a GitHub Actions workflow so that we can run tests against a new shard before distributing the updated TUF metadata.
* Turn down/freezing. We haven't gone through this process yet, so I'm sure there will be bumps. In particular, we need to figure out the order of operations for tearing down NEGs. We've already added support in Terraform for deleting the signing key and databases.

I also assume we'll update the TUF playbook part after we go through it with staging.

ramonpetgrave64 commented on SHARDING.md:
```diff
@@ -0,0 +1,451 @@
+needing to manually update the prober. The prober should pick up new shards based on the
+SigningConfig (for the write path) and TrustedRoot (for the read path).
+
+We will need a mechanism for running prober tests against a new deployment before we've
```
Also, v2 shards aren't directly discoverable from the trusted root, because the trusted root does not contain `apiMajorVersion`, like the signing config does. We can work around this by querying `<shard url>/api/v2/checkkpoint`, though.

haydentherapper replied:
I updated this to specify that you should use the SigningConfig for shard discovery, as it will contain both the active and all inactive shards. If a prober needs to correlate between the SigningConfig and TrustedRoot, it can use the shard URL.

haydentherapper commented on SHARDING.md:
```diff
@@ -0,0 +1,451 @@
+Make a post on Slack letting the community know we'll be freezing
+and turning down the log.
+
+### Disable any alerts
+
+Make sure to either disable or remove via Terraform any alerts
+for failing healthchecks or missing metrics. As we turn down the service
+and delete resources, we expect that healthchecks will start to fail.
```
Changed this section to a WIP since we don't have the list of alerts yet.

haydentherapper commented on SHARDING.md:
```diff
@@ -0,0 +1,451 @@
+
+## Terraform Module Changes
+
+New shards are declared as additional modules under the respective
+environment. This is a different pattern than other modules, which are all
+initialized through the
+[`sigstore` module](https://github.com/sigstore/terraform-modules/tree/main/gcp/modules/sigstore).
+This design allowed deployers to initialize this one module and set all required variables, at
+which point you'd have a complete Sigstore stack on GCP. The downside of this approach is
+private deployers that don't need all GCP infrastructure have to initialize all resources,
+though deployers can always initialize each module referenced in `sigstore` individually.
+
+The other downside to referencing the `sigstore` module is that an update to a sub-module
+will concurrently update all deployed shards.  Declaring each shard as its own module allows us
+to make updates to shards one-by-one.
+
+The tiled transparency log module we'll reference in Terraform is
+[`tiles_tlog`](https://github.com/sigstore/terraform-modules/tree/main/gcp/modules/tiles_tlog).
```
This is just to give background, either seems fine.

ramonpetgrave64 commented on SHARDING.md:
```diff
@@ -0,0 +1,465 @@
+
+### Retrieve shard public key
+
+Look at the service logs to find the public key. This is the easiest way to retrieve
```
Add a link to the service logs for an existing shard, so we can easily know how to find the logs for the next shard.

haydentherapper replied:
The audience for this doc is someone familiar with the infrastructure so they should be familiar with where to retrieve logs for pods.

cmurphy commented on SHARDING.md:
```diff
@@ -0,0 +1,464 @@
+
+### Freeze the log
+
+Pass `--read-only=true` to the service, which will stop write traffic and
```
I think we could skip this step and just turn off the K8s deployment.

haydentherapper replied:
I'm good with that. The only benefit of freezing the log is to gracefully reject write traffic, but unless we're adding in a waiting period for clients to fail and update to the latest shard (which shouldn't be an issue as clients should be using TUF), this doesn't add much.

haydentherapper commented on SHARDING.md:
````diff
@@ -0,0 +1,464 @@
+5. Using [tinkey](https://developers.google.com/tink/tinkey-overview#installation), create an encrypted Tink keyset:
+
+```
+tinkey create-keyset --key-template ED25519 --out enc-keyset.cfg --master-key-uri gcp-kms://projects/<project name>/locations/<region>/keyRings/<key-ring>/cryptoKeys/checkpoint-signer-key-encryption-key
````
For the 2025-1 shard, [the key name variable](https://github.com/sigstore/public-good-instance/blob/3c97f5b6b6f52d2eb38018619555c57026241a67/terraform/environments/staging/1-infrastructure/staging.tf#L242) is set to `checkpoint-signer`. For future shards, we can just omit the key name variable and the use the new default `checkpoint-signer-key-encryption-key`.

haydentherapper commented on SHARDING.md:
````diff
@@ -0,0 +1,464 @@
+
+```
+resource "google_kms_key_ring_iam_member" "rekor-tiles-keyring" {
+  key_ring_id = "<project-rekor or projectsigstore-staging>/<location, e.g. global>/<name of shard keyring>"
````
Thanks, I added how the string is constructed.

haydentherapper commented on SHARDING.md:
```diff
@@ -0,0 +1,464 @@
+
+First, remove the protection bit on the databases. In the Terraform configuration,
+set `spanner_database_sequencer_deletion_protection` and `spanner_database_antispam_deletion_protection`
+to `true`. Create a PR, merge, `plan`, and `apply`.
```
Whoops, thanks.

haydentherapper commented on SHARDING.md:
```diff
@@ -0,0 +1,464 @@
+of one week, this means setting the validity window to be at least one week
+past when the TUF targets are signed.
+
+You will need two timestamps:
```
Thanks, updated.

haydentherapper commented:
Will merge and iterate on this after we practice this playbook.
